src/vcf/validator.cpp

```diff
@@ -193,26 +208,38 @@ namespace ebi
             << " compression)";
     }
 
-    bool is_compressed_magic_num(std::istream &input)
+    std::string is_compressed_magic_num(std::istream &input)
     {
         unsigned char magic[9];
         input.read((char*)magic, sizeof(magic));
         input.seekg(0);
 
         std::vector<std::pair<std::vector<unsigned char>, std::string>> types = {
-            { { 80, 75, 3, 4 }, ZIP },
+            { { 66, 90, 104 }, BZ2 },
             { { 31, 139 }, TAR_GZ },
             { { 253, 55, 122, 88, 90 }, TAR_XZ },
-            { { 31, 157}, TAR_Z }
+            { { 31, 157}, TAR_Z },
+            { { 80, 75, 3, 4 }, ZIP }
         };
 
         for (auto & type : types) {
             if (std::equal(type.first.begin(), type.first.end(), magic)) {
                 compressed_file_warning(type.second);
-                return true;
+                return type.second;
             }
         }
-        return false;
+        return NO_EXT;
+    }
+
+    void read_compressed(std::istream & input,
+                         std::stringstream & strr,
+                         std::string const & file_extension)
+    {
+        boost::iostreams::filtering_istream in;
+        in.push(boost::iostreams::bzip2_decompressor());
+        in.push(input);
+        boost::iostreams::copy(in, strr);
+        input.rdbuf(strr.rdbuf());
```

Comment:
cpp references says that its pointless to make a copy of a stream. so what i am doing here is making a new stream of decompressed input and then giving its buffer to original input stream .
but as you told that files may be large , 
this approach needs to be dropped. 
I will figure think something to work in junks or line by line.

Reply:
Are you sure it's copying the whole buffer? I thought that at first, but then I saw that `rdbuf` just sets a pointer to the buffer, which in my understanding can be empty or not read yet. I wouldn't discard this approach yet

Reply:
I think line 241 `boost::iostreams::copy(in, strr);` will be having issue while copying large file. I will research over it.

Reply:
@jmmut just I am asking for curiosity and knoweledge. what may be the maximum size of file we will be dealing with. I just want to know the extreme limit.

Reply:
A VCF is basically a matrix, one dimension is region in the genome, and the other dimension is the samples (individuals) each cell in the matrix is "what variant does a given sample have in a given genomic position". There's a limit on how many variants you can put, because any species DNA is finite, but in theory you can put an arbitrary number of samples in the same file.

Some big files we have at EVA may be well around 30BG, if I remember correctly, but there's no limit. Some other projects are handling VCFs of hundreds of GB or even a few TB, because they join 10^4 to 10^5 samples in the same file.

We had a bug some time ago that we detected validating a file of 3 billion lines. It was most likely an integer overflow (past 2147 million), just like youtube with gangnam style.

Reply:
that's why it's so important the memory usage. We put some effort to be sure that the memory usage in the validator is constant, so it doesn't increase with time. That way, the file can be arbitrarily big.

Reply:
@jmmut I wanted to discuss something regarding a issue mentioned in gsoc projects. The issue regarding "users can validate their local files only" .
I sent a mail regarding my ideas that can be implemented as far as I understood the problem. Please have a look at them and tell me if I am going in right direction.

Reply:
as the file may be large. I was thinking it to break it into chunks before transferring the file over network. Detailed idea is in the link attached in mail. 
Also I was facing a problem regarding some communication. As I am not able to discuss ideas and doubts properly due to lack of a communication channel. I wanted to know if there is no issue in getting my doubts cleared over mail ?

Reply:
Yes, there's no problem with talking through email, I'll answer you there.

Reply:
@jmmut  there is some issue while reading compressed file.
the `readline` function seems to read multiple lines instead of single line (as it may not get '\n' in compressed format there after the first line) in case of a compressed file.
I have tested it on several test files. it seems that it reads whole of the file in one go.
let us assume it reads whole of the file in one go if it is a compressed file.
So, it is not good behaviour of `readline` if it does so. so we should need to modify` readline` function a bit (if possible).